File: apps/retailer/lib/vision/detectors/HybridDetector.js
import { LOCAL_CONFIDENCE_THRESHOLD } from '../types';
import { LocalDetector } from './LocalDetector';
import { ServerDetector } from './ServerDetector';

/**
 * Orquestrador híbrido: offline primeiro, nuvem se confiança baixa.
 */
export class HybridDetector {
  /**
   * @param {{ local?: import('./LocalDetector').LocalDetector, server?: import('./ServerDetector').ServerDetector, threshold?: number }} [options]
   */
  constructor(options = {}) {
    this.local = options.local || new LocalDetector();
    this.server = options.server || new ServerDetector();
    this.threshold = options.threshold ?? LOCAL_CONFIDENCE_THRESHOLD;
  }

  async init() {
    await this.local.load();
  }

  /**
   * @param {HTMLCanvasElement} frame
   * @returns {Promise<import('../types').VisionDetection & { usedServer: boolean }>}
   */
  async detect(frame) {
    const localResult = await this.local.detect(frame);

    if (localResult.confidence >= this.threshold && localResult.label !== 'unknown') {
      return { ...localResult, usedServer: false };
    }

    const serverResult = await this.server.detect(frame);
    return { ...serverResult, usedServer: true };
  }

  async dispose() {
    await this.local.dispose();
  }
}

File: apps/retailer-android/app/src/main/java/br/com/finmemory/retailer/painel/vision/HybridDetector.kt
package br.com.finmemory.retailer.painel.vision

class HybridDetector(
    private val local: LocalDetector = LocalDetector(),
    private val server: ServerDetector = ServerDetector(),
    private val threshold: Float = LOCAL_CONFIDENCE_THRESHOLD,
) : DetectorService {

    suspend fun init() {
        local.load()
    }

    override suspend fun isReady(): Boolean = local.isReady()

    override suspend fun detect(frame: ByteArray, width: Int, height: Int): VisionDetection {
        val localResult = local.detect(frame, width, height)
        if (localResult.confidence >= threshold && localResult.label != "unknown") {
            return localResult.copy(usedServer = false)
        }
        return server.detect(frame, width, height)
    }

    fun dispose() {
        local.dispose()
    }
}
